File: apps/android/app/src/main/kotlin/com/loonext/android/features/thread/ThreadQuotes.kt
package com.loonext.android.features.thread

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Send
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.loonext.android.AppGraph
import com.loonext.android.core.data.CacheKeys
import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.features.quotes.QUOTE_STATUS_KEYS
import com.loonext.android.features.quotes.Quote
import com.loonext.android.features.quotes.QuoteStatus
import com.loonext.android.features.quotes.Quotes
import com.loonext.android.features.quotes.QuotesRepository
import com.loonext.android.features.settings.formatMoney
import com.loonext.android.ui.common.LoadState
import com.loonext.android.ui.common.rememberCacheFirst
import com.loonext.android.ui.common.rememberHaptics
import com.loonext.android.ui.common.userMessage
import kotlinx.coroutines.launch

/**
 * #287 — what this thread has been quoted, and the way to quote it.
 *
 * ## Why a strip beside the composer rather than a bubble
 *
 * The same reasoning #224's payment strip settled, and it applies harder here.
 * The message carrying the quote link is already in the transcript exactly as
 * the customer received it. What is NOT in the transcript is the STATE — sent,
 * opened, accepted, lapsed — and three of those four change with nobody in the
 * workspace doing anything. A bubble would have to mutate after the fact, which
 * is the one thing a transcript must never do.
 *
 * ## The status it renders
 *
 * [Quote.shownStatus], never the stored column and never the wire's
 * `effective_status`. Nothing writes `expired`, and a derived string is only as
 * fresh as the read that brought it — see QuotesData.kt.
 *
 * ## The form
 *
 * *Smart Defaults*: the expiry is pre-filled at 14 days. It is the one field
 * whose answer a crew member does not care about and cannot leave blank, so
 * asking them to type it is pure friction; the amount and the work are theirs
 * and are deliberately empty, because a default price is a wrong price.
 *
 * *Ethical Friction, calibrated*: creating is not sending. A draft costs
 * nothing and is invisible to the customer, so it needs no ceremony. SEND is
 * the customer-visible act that binds a price, so the button carries the amount
 * rather than saying "Send" — you cannot press it without the figure in your
 * eye.
 *
 * Mirrors apps/web/src/components/thread/quote-strip.tsx.
 */

/** Rows worth keeping above the composer: live, or decided in the last week. */
private const val RECENT_DECISION_MS = 7L * 24 * 60 * 60 * 1000

/** The expiry a crew member does not have to think about. */
private const val DEFAULT_EXPIRY_DAYS = 14L

private fun isWorthShowing(quote: Quote, now: Long): Boolean {
    if (!Quotes.isDecided(quote.shownStatus(now))) return true
    val decided = quote.decided_at?.let {
        runCatching { java.time.Instant.parse(it).toEpochMilli() }.getOrNull()
    } ?: return false
    return now - decided < RECENT_DECISION_MS
}

@Composable
fun ThreadQuotes(
    graph: AppGraph,
    companyId: String,
    conversationId: String,
    /**
     * May this reader act? A view-only observer (#315) still SEES what the
     * thread was quoted — that is a fact about the conversation they are here
     * to read — and is offered no way to add or send one.
     */
    canAct: Boolean,
    onNotice: (String) -> Unit,
    /** A sent quote goes out as an ordinary text the timeline does not know of. */
    onSent: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val repo = remember(graph) { QuotesRepository(graph.api) }
    var refreshKey by remember { mutableIntStateOf(0) }
    val state = rememberCacheFirst(
        cache = graph.storeCache,
        key = CacheKeys.quotes(companyId, conversationId),
        refreshKey = refreshKey,
    ) { repo.forConversation(companyId, conversationId).data }

    val rows = (state as? LoadState.Ready)?.value ?: emptyList()

    val coroutines = rememberCoroutineScope()
    val haptics = rememberHaptics()
    val locale = LocalAppLocale.current
    var open by remember { mutableStateOf(false) }
    var amount by remember { mutableStateOf("") }
    var description by remember { mutableStateOf("") }
    var sendingId by remember { mutableStateOf<String?>(null) }
    var busy by remember { mutableStateOf(false) }

    val now = System.currentTimeMillis()
    val shown = rows.filter { isWorthShowing(it, now) }

    // Absent entirely when there is nothing to say and nothing to offer. A
    // strip reserving space on every thread for something almost no thread has
    // is a permanent cost paid for a rare event.
    if (shown.isEmpty() && !canAct) return

    fun create() {
        val dollars = amount.replace(Regex("[^0-9.]"), "").toDoubleOrNull()
        if (dollars == null || dollars <= 0) {
            onNotice(AppStrings.translate(locale, "quotes.needAmount"))
            return
        }
        if (description.isBlank()) {
            onNotice(AppStrings.translate(locale, "quotes.needDescription"))
            return
        }
        busy = true
        coroutines.launch {
            try {
                repo.create(
                    companyId = companyId,
                    conversationId = conversationId,
                    amountCents = Math.round(dollars * 100).toInt(),
                    description = description.trim(),
                    expiresAt = java.time.Instant
                        .ofEpochMilli(now + DEFAULT_EXPIRY_DAYS * 24 * 60 * 60 * 1000)
                        .toString(),
                )
                amount = ""
                description = ""
                open = false
                refreshKey++
            } catch (cause: Exception) {
                onNotice(cause.userMessage(locale))
            } finally {
                busy = false
            }
        }
    }

    fun send(id: String) {
        sendingId = id
        coroutines.launch {
            try {
                repo.send(companyId, id)
                haptics.confirm()
                refreshKey++
                onSent()
            } catch (cause: Exception) {
                onNotice(cause.userMessage(locale))
            } finally {
                sendingId = null
            }
        }
    }

    Column(modifier.fillMaxWidth().padding(horizontal = 4.dp)) {
        shown.forEach { quote ->
            QuoteRow(
                quote = quote,
                now = now,
                canSend = canAct,
                sending = sendingId == quote.id,
                onSend = { send(quote.id) },
            )
        }

        if (canAct && !open) {
            TextButton(onClick = { open = true }) {
                Icon(
                    Icons.Outlined.Description,
                    contentDescription = null,
                    modifier = Modifier.padding(end = 4.dp),
                )
                Text(t("quotes.newQuote"), fontSize = 13.sp)
            }
        }

        if (canAct && open) {
            Row(
                Modifier.fillMaxWidth().padding(top = 4.dp),
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                OutlinedTextField(
                    value = amount,
                    onValueChange = { amount = it },
                    label = { Text(t("quotes.amountLabel")) },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
                    modifier = Modifier.weight(0.4f),
                )
                OutlinedTextField(
                    value = description,
                    onValueChange = { description = it },
                    label = { Text(t("quotes.descriptionLabel")) },
                    singleLine = true,
                    modifier = Modifier.weight(0.6f),
                )
            }
            // The default, said out loud rather than hidden in a field nobody
            // filled in. A price with no expiry binds the business forever.
            Text(
                t("quotes.expiresInDays", "days" to DEFAULT_EXPIRY_DAYS.toString()),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 4.dp),
            )
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                TextButton(onClick = ::create, enabled = !busy) {
                    Text(
                        if (busy) t("quotes.saving") else t("quotes.saveDraft"),
                        fontSize = 13.sp,
                    )
                }
                TextButton(onClick = { open = false }) {
                    Text(t("common.cancel"), fontSize = 13.sp)
                }
            }
        }
    }
}

@Composable
private fun QuoteRow(
    quote: Quote,
    now: Long,
    canSend: Boolean,
    sending: Boolean,
    onSend: () -> Unit,
) {
    val status = quote.shownStatus(now)
    // The row carries its own currency, the way a payment request does: a quote
    // is denominated when it is written, and a workspace that later changes
    // billing currency must not restate old prices.
    val money = formatMoney(quote.amount_cents, quote.money)

    Row(
        Modifier.fillMaxWidth().padding(vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Icon(
            Icons.Outlined.Description,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(money, fontSize = 13.sp, fontWeight = FontWeight.Medium)
        Text(
            quote.description,
            fontSize = 13.sp,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f),
        )
        Text(
            t(QUOTE_STATUS_KEYS[status] ?: "quotes.statusDraft"),
            fontSize = 13.sp,
            color = when (status) {
                QuoteStatus.ACCEPTED -> MaterialTheme.colorScheme.primary
                QuoteStatus.EXPIRED, QuoteStatus.DECLINED ->
                    MaterialTheme.colorScheme.onSurfaceVariant
                else -> MaterialTheme.colorScheme.onSurface
            },
        )
        if (status == QuoteStatus.DRAFT && canSend) {
            // The amount rides on the button: this is the act the customer
            // sees, and it binds a price.
            TextButton(onClick = onSend, enabled = !sending) {
                Icon(
                    Icons.Outlined.Send,
                    contentDescription = null,
                    modifier = Modifier.padding(end = 4.dp),
                )
                Text(
                    if (sending) t("quotes.sending") else t("quotes.sendFor", "amount" to money),
                    fontSize = 13.sp,
                )
            }
        }
    }
}

File: apps/ios/Loonext/Features/Quotes/ThreadQuotes.swift
import SwiftUI

/// #287 — what this thread has been quoted, and the way to quote it.
///
/// ## Why a strip beside the composer rather than a bubble
///
/// The same reasoning #224's payment strip settled, and it applies harder here.
/// The message carrying the quote link is already in the transcript exactly as
/// the customer received it. What is NOT in the transcript is the STATE — sent,
/// opened, accepted, lapsed — and three of those four change with nobody in the
/// workspace doing anything. A bubble would have to mutate after the fact, which
/// is the one thing a transcript must never do.
///
/// ## The status it renders
///
/// `Quote.shownStatus`, never the stored column and never the wire's
/// `effective_status` — see `QuotesApi.swift` for why a cached row outlives its
/// own derivation.
///
/// ## The form
///
/// *Smart Defaults*: the expiry is pre-filled at 14 days. It is the one field
/// whose answer a crew member does not care about and cannot leave blank, so
/// asking them to type it is pure friction; the amount and the work are theirs
/// and are deliberately empty, because a default price is a wrong price.
///
/// *Ethical Friction, calibrated*: creating is not sending. A draft costs
/// nothing and is invisible to the customer, so it needs no ceremony. SEND is
/// the customer-visible act that binds a price, so the button carries the amount
/// rather than saying "Send" — you cannot press it without the figure in your
/// eye.
///
/// Mirrors `apps/web/src/components/thread/quote-strip.tsx` and
/// `ThreadQuotes.kt`.

/// Rows worth keeping above the composer: live, or decided in the last week.
private let quoteRecentDecisionSeconds: TimeInterval = 7 * 24 * 60 * 60

/// The expiry a crew member does not have to think about.
private let quoteDefaultExpiryDays = 14

private func quoteIsWorthShowing(_ quote: Quote, now: Date) -> Bool {
    if !Quotes.isDecided(quote.shownStatus(now: now)) { return true }
    guard let decided = Quotes.isoDate(quote.decided_at) else { return false }
    return now.timeIntervalSince(decided) < quoteRecentDecisionSeconds
}

@MainActor
@Observable
final class ThreadQuotesModel {
    private let quotes: QuotesApi
    private let companyId: String
    private let conversationId: String

    var rows: [Quote] = []
    var sendingId: String?
    var busy = false
    var notice: String?

    init(quotes: QuotesApi, companyId: String, conversationId: String) {
        self.quotes = quotes
        self.companyId = companyId
        self.conversationId = conversationId
    }

    var visible: [Quote] {
        let now = Date()
        return rows.filter { quoteIsWorthShowing($0, now: now) }
    }

    func load() async {
        // A failed read is silence, not an error banner. The strip is absent on
        // almost every thread anyway, and a red row above the composer for a
        // list that is usually empty is a worse trade than showing nothing.
        rows = (try? await quotes.forConversation(
            companyId: companyId,
            conversationId: conversationId
        ))?.data ?? []
    }

    func create(amountCents: Int, description: String, locale: String?) async {
        busy = true
        defer { busy = false }
        let expires = ISO8601DateFormatter().string(
            from: Date().addingTimeInterval(
                TimeInterval(quoteDefaultExpiryDays * 24 * 60 * 60)
            )
        )
        do {
            _ = try await quotes.create(
                companyId: companyId,
                conversationId: conversationId,
                amountCents: amountCents,
                description: description,
                expiresAt: expires
            )
            await load()
        } catch {
            notice = error.userMessage(locale ?? MessageLocale.en)
        }
    }

    func send(_ quote: Quote, locale: String?) async -> Bool {
        sendingId = quote.id
        defer { sendingId = nil }
        do {
            _ = try await quotes.send(companyId: companyId, quoteId: quote.id)
            await load()
            return true
        } catch {
            notice = error.userMessage(locale ?? MessageLocale.en)
            return false
        }
    }
}

struct ThreadQuotesPane: View {
    let api: ApiClient
    let companyId: String
    let conversationId: String
    let role: String?
    /// #106: 'text' or 'note' on this conversation's number.
    let viewerLevel: String
    /// A sent quote goes out as an ordinary text the transcript does not know of.
    let onSent: () -> Void

    @Environment(\.appLocale) private var appLocale

    @State private var model: ThreadQuotesModel?
    @State private var open = false
    @State private var amount = ""
    @State private var work = ""

    private func t(_ key: String, _ vars: [String: String] = [:]) -> String {
        AppStrings.translate(appLocale, key, vars)
    }

    /// A notes-only viewer sees none of this. A note goes to the crew; a quote
    /// goes to the customer, and somebody who cannot text this customer has no
    /// business quoting them.
    private var canSee: Bool { viewerLevel == "text" }

    /// #315: a view-only observer READS the thread and changes nothing in it.
    /// They still see what was quoted — that is a fact about the conversation
    /// they are here to read — and are offered no way to add or send one.
    private var canAct: Bool { canSee && role != MemberRole.readOnly }

    var body: some View {
        Group {
            if canSee, let model, !model.visible.isEmpty || canAct {
                VStack(alignment: .leading, spacing: 4) {
                    ForEach(model.visible) { quote in
                        QuoteStripRow(
                            quote: quote,
                            canSend: canAct,
                            sending: model.sendingId == quote.id
                        ) {
                            Task {
                                if await model.send(quote, locale: appLocale) { onSent() }
                            }
                        }
                    }
                    if canAct {
                        if open {
                            form(model)
                        } else {
                            Button(t("quotes.newQuote")) { open = true }
                                .font(.golos(13))
                                .foregroundStyle(BrandColor.olive)
                        }
                    }
                    // The API's own words on a refusal, never a paraphrase, and
                    // never swallowed: a create or send that failed silently is
                    // a crew member who thinks a customer has a price.
                    InlineError(model.notice)
                }
                .padding(.horizontal, 12)
                .padding(.bottom, 4)
            }
        }
        .task(id: "\(companyId)|\(conversationId)") {
            let created = ThreadQuotesModel(
                quotes: QuotesApi(api: api),
                companyId: companyId,
                conversationId: conversationId
            )
            model = created
            await created.load()
        }
    }

    @ViewBuilder
    private func form(_ model: ThreadQuotesModel) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 6) {
                TextField(t("quotes.amountLabel"), text: $amount)
                    .keyboardType(.decimalPad)
                    .textFieldStyle(.roundedBorder)
                    .frame(maxWidth: 120)
                TextField(t("quotes.descriptionLabel"), text: $work)
                    .textFieldStyle(.roundedBorder)
            }
            // The default, said out loud rather than hidden in a field nobody
            // filled in. A price with no expiry binds the business forever.
            Text(t("quotes.expiresInDays", ["days": String(quoteDefaultExpiryDays)]))
                .font(.golos(12))
                .foregroundStyle(BrandColor.muted500)
            HStack(spacing: 10) {
                Button(model.busy ? t("quotes.saving") : t("quotes.saveDraft")) {
                    Task { await submit(model) }
                }
                .font(.golos(13, weight: .semibold))
                .foregroundStyle(BrandColor.olive)
                .disabled(model.busy)

                Button(t("common.cancel")) { open = false }
                    .font(.golos(13))
                    .foregroundStyle(BrandColor.muted500)
            }
        }
    }

    private func submit(_ model: ThreadQuotesModel) async {
        let digits = amount.filter { $0.isNumber || $0 == "." }
        guard let dollars = Double(digits), dollars > 0 else {
            model.notice = t("quotes.needAmount")
            return
        }
        guard !work.trimmingCharacters(in: .whitespaces).isEmpty else {
            model.notice = t("quotes.needDescription")
            return
        }
        await model.create(
            amountCents: Int((dollars * 100).rounded()),
            description: work.trimmingCharacters(in: .whitespaces),
            locale: appLocale
        )
        amount = ""
        work = ""
        open = false
    }
}

private struct QuoteStripRow: View {
    let quote: Quote
    let canSend: Bool
    let sending: Bool
    let onSend: () -> Void

    @Environment(\.appLocale) private var appLocale

    var body: some View {
        let status = quote.shownStatus()
        // The row carries its own currency, the way a payment request does: a
        // quote is denominated when it is written, and a workspace that later
        // changes billing currency must not restate old prices.
        let money = quote.amountLabel

        HStack(spacing: 6) {
            Image(systemName: "doc.text")
                .foregroundStyle(BrandColor.muted500)
            Text(money)
                .font(.golos(13, weight: .medium))
            Text(quote.description)
                .font(.golos(13))
                .foregroundStyle(BrandColor.muted500)
                .lineLimit(1)
            Spacer(minLength: 8)
            Text(AppStrings.translate(appLocale, QuoteStatus.keys[status] ?? "quotes.statusDraft"))
                .font(.golos(13))
                .foregroundStyle(statusTint(status))
            if status == QuoteStatus.draft, canSend {
                // The amount rides on the button: this is the act the customer
                // sees, and it binds a price.
                Button(
                    sending
                        ? AppStrings.translate(appLocale, "quotes.sending")
                        : AppStrings.translate(appLocale, "quotes.sendFor", ["amount": money])
                ) {
                    onSend()
                }
                .font(.golos(13, weight: .semibold))
                .foregroundStyle(BrandColor.olive)
                .disabled(sending)
            }
        }
    }

    private func statusTint(_ status: String) -> Color {
        switch status {
        case QuoteStatus.accepted: return BrandColor.olive
        case QuoteStatus.expired, QuoteStatus.declined: return BrandColor.muted500
        default: return BrandColor.ink
        }
    }
}
